src/nxt_cgroup.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_process.h>
+#include <nxt_malloc.h>
+#include <nxt_fs.h>
+#include <nxt_cgroup.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <limits.h>
+
+
+static char *nxt_mk_cgpath(const char *dir);
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    int   ret, len;
+    FILE  *fp;
+    char  *cgpath;
+    char  cgprocs[PATH_MAX - 13];   /* - 13 for '/cgroup.procs' */
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.path == NULL)
+    {
+        return NXT_OK;
+    }
+
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgpath == NULL) {
+        return NXT_ERROR;
+    }
+    nxt_fs_mkdir_all((const u_char *) cgpath, 0777);
+
+    snprintf(cgprocs, sizeof(cgprocs), "%s/cgroup.procs", cgpath);
+    fp = fopen(cgprocs, "we");
+    if (fp == NULL) {
+        ret = NXT_ERROR;
+        goto out_free;
+    }
+
+    ret = NXT_OK;
+
+    setvbuf(fp, NULL, _IONBF, 0);
+    len = fprintf(fp, "%d\n", process->pid);
+    if (len < 0) {
+        ret = NXT_ERROR;
+    }
+
+    fclose(fp);
+
+out_free:
+
+    nxt_free(cgpath);
+
+    return ret;
+}
+
+
+void
+nxt_cgroup_cleanup(const nxt_process_t *process)
+{
+    char  *cgroot, *cgpath, *ptr;
+
+    cgroot = nxt_mk_cgpath("");
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgroot == NULL || cgpath == NULL) {
+        goto out_free;
+    }
+
+    while (*cgpath != '\0' && strcmp(cgroot, cgpath) != 0) {
+        rmdir(cgpath);
+        ptr = strrchr(cgpath, '/');
+        *ptr = '\0';
+    }
+
+out_free:
+
+    nxt_free(cgroot);
+    nxt_free(cgpath);
+}
+
+
+static char *
+nxt_mk_cgpath(const char *dir)
+{
+    int   len;
+    char  *path, *buf, *ptr;
+
+    path = buf = ptr = NULL;
+
+    /*
+     * If the path from the config is relative, we need to make
+     * the cgroup path include the main unit processes cgroup. I.e
+     *
+     *   NXT_CGROUP_ROOT/<main process cgroup>/<cgroup path>
+     */
+    if (dir[0] != '/') {
+        int      i;
+        bool     found;
+        FILE     *fp;
+        size_t   len;
+        ssize_t  nread;
+
+        fp = fopen("/proc/self/cgroup", "re");
+        if (fp == NULL) {
+            return NULL;
+        }
+
+        found = false;
+        while ((nread = getline(&buf, &len, fp)) != -1) {
+            if (strncmp(buf, "0::", 3) == 0) {
+                found = true;
+                break;
+            }
+        }
```

Comment:
Is there a memory leak when we read more than one line?

Reply:
Hmm, yes, I think your right, buf will be allocated each time round the loop.

Reply:
Sorry; my mistake.

I just re-read getline(3) (and getline(3POSIX)), and it seems it's designed to handle this nicely: it reuses the buffer and realloc(3)s as necessary.

Reply:
`getline(3)` gets an input buffer by a non-`NULL` pointer and the size.  So the code you were using originally was actually good (since by not modifying the size, it kept the old value, and the pointer was of course the same).  That's the way for `getline(3)` to know about the buffer.  So it seems it doesn't care if it allocated in a previous run, or if you allocated through `malloc(3)`.

But yeah, a wee test case + valgrind seems the right thing to do :)

Reply:
Yes, you're right, confirmed with valgrind.

If we don't NULL buf then getline(3) will re-use it.

OK, cool, I'll put it back to how it was.